add C-based `isAscii :: Text -> Bool`

raehik opened the pull request:
Now that `Text`s are represented internally as a UTF-8 bytestring, we can provide a fast `isAscii :: Text -> Bool` that inspects the bytestring directly, rather than `Char` by `Char`. Such a function can come in handy in serialization libraries, and can't easily be written by an end user. Plus, the C snippet already exists, so there's not much to do.

raehik commented:
I am almost certain my FFI code is buggy. I don't well understand pinned/unpinned memory. But I'm unable to test on my machine (build issues).

phadej commented on src/Data/Text.hs:
```diff
@@ -1176,6 +1180,21 @@ minimum :: HasCallStack => Text -> Char
+-- TODO
+isAscii :: Text -> Bool
+isAscii (Text (A.ByteArray arr) off len) =
+    let arrPtr = Exts.Ptr (Exts.byteArrayContents# arr)
```
This is unsafe/wrong. `Text`'s bytearray is not pinned, so it may move after you take a pointer. `ccall unsafe` is not enough.

You need to import `_hs_text_is_ascii` as taking `ByteArray#` argument (grep for `memchr`). Then there won't be a race condition.

raehik replied:
I thought that was wrong! Thanks very much for the help, I don't understand memory pinning & had hoped to get poked in the right place. So the FFI resolves `ByteArray#` and `Ptr Word8` to the same C type, but they do different things. Is this hidden in the compiler somewhere? *Edit: I failed to read a line in GHC.Exts about `UnliftedFFITypes`! There's even a table about usage safety! Awesome.*

If you don't mind, another thing I was couldn't figure out: what is the difference between an `IO a` and a "pure" `a` foreign import? (Either way it's the same C return type, right? just info for the compiler?)

phadej replied:
If the function is pure like `sin` then you don't need `IO`.

In

```haskell
foreign import ccall unsafe "_hs_text_is_ascii" c_is_ascii
    :: Ptr Word8 -> Ptr Word8 -> IO CSize
```

The C function reads memory, so that is not pure. (Another thread can write into the same memory range e.g.)

```haskell
foreign import ccall unsafe "_hs_text_is_ascii_2" c_is_ascii_2
    :: ByteArray# -> CSize -> IO CSize
```

is OTOH "pure" as `ByteArray#` is passed in "by value". That's why `IO` was dropped recently (See https://github.com/haskell/text/pull/482)

raehik replied:
I see! I've done your 2nd example without the `IO`, which calls the main function with pointers to the beginning and end of the data calculated (the same pattern is used in some `Data.Text.Array` foreign calls):

```haskell
isAscii (Text (A.ByteArray arr) (Exts.I# off#) (Exts.I# len#)) =
    cSizeToInt (c_is_ascii_2 arr off# len#) == (Exts.I# len#)
foreign import ccall unsafe "_hs_text_is_ascii_2" c_is_ascii_2
    :: ByteArray# -> Exts.Int# -> Exts.Int# -> CSize
```

phadej commented on src/Data/Text.hs:
```diff
@@ -1176,6 +1179,19 @@ minimum :: HasCallStack => Text -> Char
+{-# INLINE cSizeToInt #-}
+
+foreign import ccall unsafe "_hs_text_is_ascii_2" c_is_ascii_2
+    :: ByteArray# -> Exts.Int# -> Exts.Int# -> CSize
```
you don't need to use `Int#` here. `CSize` works fine (and is actually more correct).

raehik replied:
OK, thanks

phadej commented on src/Data/Text.hs:
```diff
@@ -264,6 +265,8 @@ import Data.Text.Foreign (asForeignPtr)
 import System.IO.Unsafe (unsafePerformIO)
 #endif
 
+import Foreign.Ptr ( plusPtr )
```
unused import?

raehik replied:
From previous definition, removed

phadej reviewed (commented):
... and you probably want to add tests.

raehik commented:
Fixed my build issues, will do. Thanks a ton for the help with the fiddly bits.

raehik commented:
Fixed an error in the C code (end of text data is at `arr+off+len`, not `arr+len`), added lazy version, added property tests for both. Ready for review/squish/merge.

raehik commented:
Something to note: `Data.Text.isAscii` clashes with `Data.Char.isAscii`. There were already a couple of clashing functions (found  `toUpper`, `toLower` on a glance), so I wager that it shouldn't be considered an issue.

Bodigrim commented on cbits/is_ascii.c:
```diff
@@ -45,3 +45,10 @@ const size_t _hs_text_is_ascii(const uint8_t *src0, const uint8_t *srcend){
+/*
+  _hs_text_is_ascii_wrapper is a helper for calling _hs_text_is_ascii on Texts.
+*/
+const size_t _hs_text_is_ascii_2(const uint8_t *arr, size_t off, size_t len){
```
```suggestion
const size_t _hs_text_is_ascii_offset(const uint8_t *arr, size_t off, size_t len){
```

raehik replied:
thanks, didn't know what to call this

Lysxia reviewed (commented):
That looks good to me!

Another pair of eyes (@phadej?) to go over the FFI bits again would be nice. 
There's the `@since` TODOs, I think this can make it into 2.0.2.

Boarders commented:
Looks good to me, would be nice to have a test for the edge cases to make sure the offset is working as expected i.e. make sure this fails on text which fails to be ascii on the boundary with non-zero offset (looks correct, but it is useful to have such tests regardless).

Bodigrim commented on src/Data/Text.hs:
```diff
@@ -1176,6 +1177,35 @@ minimum :: HasCallStack => Text -> Char
+--
+-- prop> isAscii t == all (< '\x80') t
+--
+-- @since 2.0.2 (?) TODO
```
Yeah, let's commit to 2.0.2 here and below.

raehik replied:
done

raehik commented:
> Looks good to me, would be nice to have a test for the edge cases to make sure the offset is working as expected i.e. make sure this fails on text which fails to be ascii on the boundary with non-zero offset (looks correct, but it is useful to have such tests regardless).

OK. I added a single test for that:

```haskell
isAscii_border :: IO ()
isAscii_border = do
    -- ASCII prefix ends at position 3 (from 0)
    let text  = T.pack "123一二三"
        text' = case text of T.Text arr off _len -> T.Text arr (off+1) 3
    assertBool "UTF-8 string with ASCII prefix ending at last position incorrectly detected as ASCII" $ not $ T.isAscii text'
```

We're really testing the behaviour of `_hs_text_is_ascii` rather than `isAscii` or the C wrapper, which do very little (a non-zero offset means very little to the C call). What do you think @Boarders ?

Lysxia commented on tests/Tests/Properties/Folds.hs:
```diff
@@ -190,6 +194,13 @@ tl_unfoldrN n m   = (L.take i . L.unfoldr (unf j)) `eq`
+    -- ASCII prefix ends at position 3 (from 0)
+    let text  = T.pack "123一二三"
+        text' = case text of T.Text arr off _len -> T.Text arr (off+1) 3
+    assertBool "UTF-8 string with ASCII prefix ending at last position incorrectly detected as ASCII" $ not $ T.isAscii text'
```
Instead of pattern-matching and constructing an invalid `Text`, you can use `T.drop` to safely get a non-zero offset. And then the ascii prefix should be longer than the byte-length after `T.drop` to properly test against accidentally ignoring the offset.

raehik replied:
thanks, made those changes

Bodigrim commented on src/Data/Text/Lazy.hs:
```diff
@@ -870,6 +871,27 @@ minimum :: HasCallStack => Text -> Char
+-- @since 2.0.2
+isAscii :: Text -> Bool
+isAscii Empty = True
+isAscii (Chunk c cs) = T.isAscii c && isAscii cs
```
Please use `foldrChunks`.

Bodigrim commented on tests/Tests/Properties/Folds.hs:
```diff
@@ -11,14 +11,16 @@ import Control.Arrow (second)
 import qualified Data.Text.Internal.Fusion.Common as S
 import qualified Data.Text.Lazy as TL
+import qualified Data.Char as Char
+import qualified Data.Text.Internal as T
```
Redundant import.

Bodigrim commented:
Thanks!
